File: src/backend/workers/parser/handler.py
"""Parser worker handler — downloads artifacts, parses them, stores results."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

import structlog
from ulid import ULID

from domains.artifacts.models import ArtifactError, ArtifactStatus
from domains.artifacts.repository import ArtifactRepository
from shared.blob import BlobService
from shared.cosmos import CosmosService
from shared.event_types import EVENT_ARTIFACT_PARSE_FAILED, EVENT_ARTIFACT_PARSED
from shared.events import EventGridPublisher, build_cloud_event
from workers.base import PermanentError, TransientError, WorkerHandler
from workers.parser.parsers import UnsupportedArtifactType, get_parser

logger: structlog.stdlib.BoundLogger = structlog.get_logger(__name__)

DATABASE_NAME = "integration-copilot"
CONTAINER_NAME = "projects"

# Statuses that indicate the artifact has already progressed past the parse stage.
_POST_PARSE_STATUSES: frozenset[ArtifactStatus] = frozenset(
    {
        ArtifactStatus.PARSED,
        ArtifactStatus.PARSE_FAILED,
        ArtifactStatus.GRAPH_BUILDING,
        ArtifactStatus.GRAPH_BUILT,
        ArtifactStatus.GRAPH_FAILED,
    }
)


class ParserHandler(WorkerHandler):
    """Process ``ArtifactScanPassed`` events by parsing the raw artifact
    and storing a structured parse result in Cosmos DB.
    """

    def __init__(
        self,
        artifact_repository: ArtifactRepository,
        blob_service: BlobService,
        cosmos_service: CosmosService,
        event_publisher: EventGridPublisher,
    ) -> None:
        self._repo = artifact_repository
        self._blob = blob_service
        self._cosmos = cosmos_service
        self._publisher = event_publisher

    # -- WorkerHandler interface ----------------------------------------------

    async def is_already_processed(self, event_data: dict[str, Any]) -> bool:
        """Return ``True`` if the artifact has already been parsed or moved past parsing."""
        tenant_id = event_data["tenantId"]
        artifact_id = event_data["artifactId"]

        artifact = await self._repo.get_by_id(tenant_id, artifact_id)
        if artifact is None:
            return False

        return artifact.status in _POST_PARSE_STATUSES

    async def handle(self, event_data: dict[str, Any]) -> None:
        """Parse the artifact: scan_passed → parsing → parsed."""
        tenant_id = event_data["tenantId"]
        project_id = event_data["projectId"]
        artifact_id = event_data["artifactId"]

        log = logger.bind(tenant_id=tenant_id, artifact_id=artifact_id)

        # Fetch artifact and transition scan_passed → parsing.
        # If the artifact is already in PARSING (retry after a transient failure),
        # skip the status transition and continue with the parse attempt.
        try:
            artifact = await self._repo.get_by_id(tenant_id, artifact_id)
        except Exception as exc:
            raise TransientError(f"Failed to fetch artifact: {exc}") from exc

        if artifact is None:
            raise PermanentError(f"Artifact {artifact_id} not found for tenant {tenant_id}")

        if artifact.status != ArtifactStatus.PARSING:
            try:
                artifact = await self._repo.update_status(tenant_id, artifact_id, ArtifactStatus.PARSING)
            except Exception as exc:
                raise TransientError(f"Failed to transition to parsing: {exc}") from exc

            if artifact is None:
                raise PermanentError(f"Artifact {artifact_id} not found for tenant {tenant_id}")

        # Use artifact record as source of truth for blob_path and artifact_type,
        # since upstream events may not include these fields.
        blob_path = artifact.blob_path
        artifact_type = artifact.artifact_type

        if not blob_path:
            raise PermanentError(f"Artifact {artifact_id} has no blob path")

        if not artifact_type:
            raise PermanentError(f"Artifact {artifact_id} has no artifact type")

        log = log.bind(artifact_type=artifact_type)

        # Download raw artifact from Blob Storage
        try:
            content = await self._blob.download_blob(blob_path)
        except Exception as exc:
            raise TransientError(f"Failed to download blob {blob_path}: {exc}") from exc

        # Select parser and parse
        try:
            parser = get_parser(artifact_type)
        except UnsupportedArtifactType as exc:
            raise PermanentError(str(exc)) from exc

        filename = blob_path.rsplit("/", 1)[-1] if "/" in blob_path else blob_path
        try:
            result = parser.parse(content, filename)
        except (ValueError, KeyError, TypeError) as exc:
            raise PermanentError(f"Parse error for {filename}: {exc}") from exc

        result.artifact_id = artifact_id
        result.artifact_type = artifact_type

        # Store parse result in Cosmos DB
        parse_result_id = f"pr_{ULID()}"
        try:
            container = await self._cosmos.get_container(DATABASE_NAME, CONTAINER_NAME)
            doc = {
                "id": parse_result_id,
                "partitionKey": tenant_id,
                "type": "parse_result",
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "artifactType": artifact_type,
                "components": [c.model_dump(by_alias=True, mode="json") for c in result.components],
                "edges": [e.model_dump(by_alias=True, mode="json") for e in result.edges],
                "externalReferences": [r.model_dump(by_alias=True, mode="json") for r in result.external_references],
                "parsedAt": result.parsed_at.isoformat(),
            }
            await container.create_item(body=doc)
        except Exception as exc:
            raise TransientError(f"Failed to store parse result: {exc}") from exc

        # parsing → parsed
        try:
            await self._repo.update_status(tenant_id, artifact_id, ArtifactStatus.PARSED)
        except Exception as exc:
            raise TransientError(f"Failed to transition to parsed: {exc}") from exc

        # Publish ArtifactParsed event
        event = build_cloud_event(
            event_type=EVENT_ARTIFACT_PARSED,
            source="/integration-copilot/worker/artifact-parser",
            subject=f"tenants/{tenant_id}/projects/{project_id}/artifacts/{artifact_id}",
            data={
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "parseResultId": parse_result_id,
                "componentCount": len(result.components),
                "edgeCount": len(result.edges),
                "parsedAt": result.parsed_at.isoformat(),
            },
        )
        await self._publisher.publish_event(event)
        log.info("artifact_parsed_event_published", parse_result_id=parse_result_id)

    async def handle_failure(self, event_data: dict[str, Any], error: Exception) -> None:
        """Transition artifact to ``parse_failed`` on permanent error."""
        tenant_id = event_data["tenantId"]
        artifact_id = event_data["artifactId"]

        try:
            artifact = await self._repo.get_by_id(tenant_id, artifact_id)
            if artifact is not None and artifact.status not in _POST_PARSE_STATUSES:
                try:
                    await self._repo.update_status(tenant_id, artifact_id, ArtifactStatus.PARSE_FAILED)
                except Exception:
                    artifact.error = ArtifactError(
                        code="PARSE_FAILED",
                        message=str(error),
                        occurredAt=datetime.now(UTC),
                    )
                    await self._repo.update(artifact)

            # Publish ArtifactParseFailed event
            event = build_cloud_event(
                event_type=EVENT_ARTIFACT_PARSE_FAILED,
                source="/integration-copilot/worker/artifact-parser",
                subject=(
                    f"tenants/{tenant_id}/projects/{event_data.get('projectId', 'unknown')}"
                    f"/artifacts/{artifact_id}"
                ),
                data={
                    "tenantId": tenant_id,
                    "projectId": event_data.get("projectId"),
                    "artifactId": artifact_id,
                    "error": str(error),
                },
            )
            await self._publisher.publish_event(event)
        except Exception:
            logger.error("handle_failure_error", artifact_id=artifact_id, exc_info=True)

File: src/worker-parser-rust/src/worker/handler.rs
//! Parser worker handler — downloads artifacts, parses them, stores results.
//!
//! Mirrors the Python `workers/parser/handler.py` ParserHandler.

use chrono::Utc;
use serde_json::{json, Value};
use ulid::Ulid;

use crate::azure::blob::BlobService;
use crate::azure::cosmos::CosmosService;
use crate::azure::event_grid::publisher::{build_cloud_event, EventGridPublisher};
use crate::event_types::{EVENT_ARTIFACT_PARSE_FAILED, EVENT_ARTIFACT_PARSED};
use crate::parsers;
use crate::worker::{WorkerError, WorkerHandler};

const DATABASE_NAME: &str = "integration-copilot";
const CONTAINER_NAME: &str = "projects";

/// Statuses that indicate the artifact has already progressed past the parse stage.
const POST_PARSE_STATUSES: &[&str] = &[
    "parsed",
    "parse_failed",
    "graph_building",
    "graph_built",
    "graph_failed",
];

/// Process `ArtifactScanPassed` events by parsing the raw artifact
/// and storing a structured parse result in Cosmos DB.
pub struct ParserHandler {
    blob: BlobService,
    cosmos: CosmosService,
    publisher: EventGridPublisher,
}

impl ParserHandler {
    pub fn new(
        blob: BlobService,
        cosmos: CosmosService,
        publisher: EventGridPublisher,
    ) -> Self {
        Self {
            blob,
            cosmos,
            publisher,
        }
    }

    /// Read the artifact document from Cosmos and return its status.
    async fn get_artifact_status(
        &self,
        tenant_id: &str,
        artifact_id: &str,
    ) -> Result<Option<String>, WorkerError> {
        let doc = self
            .cosmos
            .read_item(DATABASE_NAME, CONTAINER_NAME, artifact_id, tenant_id)
            .await
            .map_err(|e| WorkerError::Transient(format!("cosmos read failed: {e}")))?;

        Ok(doc
            .and_then(|d| d.get("status").and_then(Value::as_str).map(String::from)))
    }

    /// Transition artifact status via a patch (read-modify-write).
    async fn update_status(
        &self,
        tenant_id: &str,
        artifact_id: &str,
        new_status: &str,
    ) -> Result<(), WorkerError> {
        let doc = self
            .cosmos
            .read_item(DATABASE_NAME, CONTAINER_NAME, artifact_id, tenant_id)
            .await
            .map_err(|e| WorkerError::Transient(format!("cosmos read failed: {e}")))?;

        let mut doc = doc.ok_or_else(|| {
            WorkerError::Permanent(format!("Artifact {artifact_id} not found for tenant {tenant_id}"))
        })?;

        let etag = doc.get("_etag").and_then(Value::as_str).map(String::from);

        if let Some(obj) = doc.as_object_mut() {
            obj.insert("status".into(), json!(new_status));
            obj.insert("updatedAt".into(), json!(Utc::now().to_rfc3339()));
        }

        self.cosmos
            .replace_item(
                DATABASE_NAME,
                CONTAINER_NAME,
                artifact_id,
                tenant_id,
                &doc,
                etag.as_deref(),
            )
            .await
            .map_err(|e| WorkerError::Transient(format!("cosmos replace failed: {e}")))?;

        Ok(())
    }
}

#[async_trait::async_trait]
impl WorkerHandler for ParserHandler {
    async fn is_already_processed(&self, event_data: &Value) -> Result<bool, WorkerError> {
        let tenant_id = event_data
            .get("tenantId")
            .and_then(Value::as_str)
            .unwrap_or("");
        let artifact_id = event_data
            .get("artifactId")
            .and_then(Value::as_str)
            .unwrap_or("");

        let status = self.get_artifact_status(tenant_id, artifact_id).await?;
        Ok(status
            .map(|s| POST_PARSE_STATUSES.contains(&s.as_str()))
            .unwrap_or(false))
    }

    async fn handle(&self, event_data: &Value) -> Result<(), WorkerError> {
        let tenant_id = str_field(event_data, "tenantId")?;
        let project_id = str_field(event_data, "projectId")?;
        let artifact_id = str_field(event_data, "artifactId")?;

        // Fetch artifact record from Cosmos and transition scan_passed → parsing.
        // If the artifact is already in PARSING (retry after a transient failure),
        // skip the status transition and continue with the parse attempt.
        let artifact = self
            .cosmos
            .read_item(DATABASE_NAME, CONTAINER_NAME, artifact_id, tenant_id)
            .await
            .map_err(|e| WorkerError::Transient(format!("Failed to fetch artifact: {e}")))?
            .ok_or_else(|| {
                WorkerError::Permanent(format!(
                    "Artifact {artifact_id} not found for tenant {tenant_id}"
                ))
            })?;

        let current_status = artifact
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or("");

        if current_status != "parsing" {
            self.update_status(tenant_id, artifact_id, "parsing")
                .await
                .map_err(|e| {
                    WorkerError::Transient(format!("Failed to transition to parsing: {e}"))
                })?;
        }

        // Use artifact record as source of truth for blob_path and artifact_type,
        // since upstream events may not include these fields.
        let blob_path = artifact
            .get("blobPath")
            .and_then(Value::as_str)
            .unwrap_or("");
        let artifact_type = artifact
            .get("artifactType")
            .and_then(Value::as_str)
            .unwrap_or("");

        if blob_path.is_empty() {
            return Err(WorkerError::Permanent(format!(
                "Artifact {artifact_id} has no blob path"
            )));
        }
        if artifact_type.is_empty() {
            return Err(WorkerError::Permanent(format!(
                "Artifact {artifact_id} has no artifact type"
            )));
        }

        // Download raw artifact from Blob Storage
        let content = self
            .blob
            .download_blob(blob_path)
            .await
            .map_err(|e| WorkerError::Transient(format!("Failed to download blob {blob_path}: {e}")))?;

        // Select parser
        let parser = parsers::get_parser(artifact_type).ok_or_else(|| {
            WorkerError::Permanent(format!("Unsupported artifact type: {artifact_type}"))
        })?;

        // Parse
        let filename = blob_path.rsplit('/').next().unwrap_or(blob_path);
        let result = parser
            .parse(&content, filename, artifact_id)
            .map_err(|e| WorkerError::Permanent(format!("Parse error for {filename}: {e}")))?;

        // Store parse result in Cosmos DB
        let parse_result_id = format!("pr_{}", Ulid::new());
        let doc = json!({
            "id": parse_result_id,
            "partitionKey": tenant_id,
            "type": "parse_result",
            "tenantId": tenant_id,
            "projectId": project_id,
            "artifactId": artifact_id,
            "artifactType": artifact_type,
            "components": result.components,
            "edges": result.edges,
            "externalReferences": result.external_references,
            "parsedAt": result.parsed_at.to_rfc3339(),
        });

        self.cosmos
            .create_item(DATABASE_NAME, CONTAINER_NAME, tenant_id, &doc)
            .await
            .map_err(|e| WorkerError::Transient(format!("Failed to store parse result: {e}")))?;

        // parsing → parsed
        self.update_status(tenant_id, artifact_id, "parsed")
            .await
            .map_err(|e| WorkerError::Transient(format!("Failed to transition to parsed: {e}")))?;

        // Publish ArtifactParsed event
        let event = build_cloud_event(
            EVENT_ARTIFACT_PARSED,
            "/integration-copilot/worker/artifact-parser",
            &format!("tenants/{tenant_id}/projects/{project_id}/artifacts/{artifact_id}"),
            json!({
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "parseResultId": parse_result_id,
                "componentCount": result.components.len(),
                "edgeCount": result.edges.len(),
                "parsedAt": result.parsed_at.to_rfc3339(),
            }),
        );

        if let Err(e) = self.publisher.publish_event(&event).await {
            tracing::warn!(error = %e, "artifact_parsed_event_publish_failed");
        } else {
            tracing::info!(parse_result_id = %parse_result_id, "artifact_parsed_event_published");
        }

        Ok(())
    }

    async fn handle_failure(&self, event_data: &Value, error: &str) {
        let tenant_id = event_data
            .get("tenantId")
            .and_then(Value::as_str)
            .unwrap_or("unknown");
        let artifact_id = event_data
            .get("artifactId")
            .and_then(Value::as_str)
            .unwrap_or("unknown");
        let project_id = event_data
            .get("projectId")
            .and_then(Value::as_str)
            .unwrap_or("unknown");

        // Try to transition to parse_failed
        if let Ok(Some(status)) = self.get_artifact_status(tenant_id, artifact_id).await
            && !POST_PARSE_STATUSES.contains(&status.as_str())
            && let Err(e) = self
                .update_status(tenant_id, artifact_id, "parse_failed")
                .await
        {
            tracing::error!(error = %e, "update_status_to_parse_failed_error");
        }

        // Publish ArtifactParseFailed event
        let event = build_cloud_event(
            EVENT_ARTIFACT_PARSE_FAILED,
            "/integration-copilot/worker/artifact-parser",
            &format!("tenants/{tenant_id}/projects/{project_id}/artifacts/{artifact_id}"),
            json!({
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "error": error,
            }),
        );

        if let Err(e) = self.publisher.publish_event(&event).await {
            tracing::error!(
                artifact_id = %artifact_id,
                error = %e,
                "handle_failure_event_publish_error"
            );
        }
    }
}

fn str_field<'a>(data: &'a Value, field: &str) -> Result<&'a str, WorkerError> {
    data.get(field)
        .and_then(Value::as_str)
        .ok_or_else(|| WorkerError::Permanent(format!("missing field: {field}")))
}
